File: algorithm_leetcode/src/main/java/com/huldar/java/day10/RemoveNthFromEnd.java
package com.huldar.java.day10;


/**
 * 移除链表的倒数第n个元素
 *
 * @author huldar
 * @date 2019/7/24 10:39
 */
public class RemoveNthFromEnd {

    public LinkNode removeNthFromEndDoubleLoop(LinkNode head, int nTh) {
        //思路:给链表添加一个哑巴节点,目的是为了防止极端情况
        LinkNode dummy = new LinkNode(0);
        dummy.next = head;
        //求出链表长度
        int length = 0;
        LinkNode first = head;
        while (first != null) {
            length++;
            first = first.next;
        }
        //要删除的前一个元素的位置
        length -= nTh;

        first = dummy;
        while (length > 0) {
            length--;
            //找到要删除的前一个元素
            first = first.next;
        }
        //将删除的前一个元素的下一个赋值给下下一个
        first.next = first.next.next;

        return dummy.next;
    }

    public LinkNode removeNthFromEndOneLoop(LinkNode head, int n) {
        //安全节点 避免极端情况
        LinkNode dummy = new LinkNode(0);
        dummy.next = head;
        //两个指针
        LinkNode first = dummy;
        LinkNode second = dummy;
        //将第一个指针指向第n+1个元素,使得两个指针之间的具体为n
        for (int i = 0; i <= n + 1; i++) {
            first = first.next;
        }
        //当第一个指针指向最后一个元素时候 第二个指针正好指向第n-1的元素
        while (first != null) {
            first = first.next;
            second = second.next;
        }
        //替换下一个元素
        second.next = second.next.next;
        return dummy.next;
    }

    class LinkNode {
        int value;
        LinkNode next;

        public LinkNode(int value) {
            this.value = value;
        }
    }
}

File: algorithm_leetcode/src/main/python/com/huldar/python/day10/RemoveNthFromEnd.py
#!usr/bin/env python
# -*- coding:utf-8 _*-
"""
@author:huldar
@file: RemoveNthFromEnd.py
@time: 2019/07/24
"""


class ListNode:
    def __init__(self, x):
        self.value = x
        self.next = None


class Solution:
    def remove_nth_from_end(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        first, second = dummy
        i = 0
        while i < n + 1:
            first = first.next
            i += 1

        while first:
            first = first.next
            second = second.next
        second.next = second.next.next
        return dummy.next

    def remove_nth_from_end_use_array(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        temp_list = []
        while dummy:
            temp_list.append(dummy)
            dummy = dummy.next
        temp_list[-n - 1].next = temp_list[-n - 1].next.next
        return temp_list[0].next


if __name__ == '__main__':
    head = ListNode(1)
    second = ListNode(2)
    three = ListNode(3)
    four = ListNode(4)
    five = ListNode(5)
    head.next = second
    second.next = three
    three.next = four
    four.next = five
    s = Solution()
    # head = s.remove_nth_from_end(head=head, n=2)
    # temp = head
    #
    # while temp:
    #     print(temp.value)
    #     temp = temp.next

    head = s.remove_nth_from_end_use_array(head=head, n=2)
    temp = head
    while temp:
        print(temp.value)
        temp = temp.next
